[SPARK-32320][PYSPARK] Remove mutable default arguments

Fokko opened the pull request:
This is bad practice, and might lead to unexpected behaviour:
https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

```
fokkodriesprong@Fan spark % grep -R "={}" python | grep def

python/pyspark/resource/profile.py:    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
python/pyspark/sql/functions.py:def from_json(col, schema, options={}):
python/pyspark/sql/functions.py:def to_json(col, options={}):
python/pyspark/sql/functions.py:def schema_of_json(json, options={}):
python/pyspark/sql/functions.py:def schema_of_csv(csv, options={}):
python/pyspark/sql/functions.py:def to_csv(col, options={}):
python/pyspark/sql/functions.py:def from_csv(col, schema, options={}):
python/pyspark/sql/avro/functions.py:def from_avro(data, jsonFormatSchema, options={}):
```

```
fokkodriesprong@Fan spark % grep -R "=\[\]" python | grep def
python/pyspark/ml/tuning.py:    def __init__(self, bestModel, avgMetrics=[], subModels=None):
python/pyspark/ml/tuning.py:    def __init__(self, bestModel, validationMetrics=[], subModels=None):
```

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?

Removing the mutable default arguments.
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?

Removing the mutable default arguments, and changing the signature to `Optional[...]`.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?

No 👍 
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?

Using the Flake8 bugbear code analysis plugin.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -2392,7 +2393,7 @@ def json_tuple(col, *fields):
 
 @since(2.1)
-def from_json(col, schema, options={}):
+def from_json(col, schema, options: Dict = None):
```
I am happy with trying Python type hints. Actually, I am thinking about somehow leveraging (or porting?) https://github.com/zero323/pyspark-stubs, cc @zero323. Let's avoid to introduce the type hinting in the main source for now because we could decouple and run it separately.

Fokko replied:
While touching the lines anyway, I thought is would be a nice addition. Having it in the source also makes live easier for the developers that work on pyspark itself. Curious what Maciej thinks.

zero323 replied:
> Furthermore, it is easier to keep the code and annotations in sync. Curious what Maciej thinks.

That's the topic for a long discussion :) Overall, I made some points [here](http://apache-spark-developers-list.1001551.n3.nabble.com/Re-PySpark-Revisiting-PySpark-type-annotations-td26232.html) and [here](http://apache-spark-developers-list.1001551.n3.nabble.com/PYTHON-PySpark-typing-hints-td21560.html) ‒ I think these are still valid.

Now... I am still opened about merging `pyspark-stubs` about with the main project. However, have some annotations in the code-base, if there is no will to maintain these (and maintaining useful annotations is honestly a pain) and bring them to high coverage any time soon, as it may make things either wasted effort (if not used at all, if package is not PEP 561 compliant) or maybe even harmful. That's a topic for a longer discussion though...

In general I'd bring the discussion to dev and or JIRA first.

Fokko replied:
Thanks for pointing out the maillist. Python <3.6 has been dropped in the meantime https://github.com/apache/spark/pull/28957.

I'm happy to help with migrating the annotations to the codebase. I still believe that it makes much sense to keep them sync with the actual code base. Otherwise you need to support multiple versions in your stubs and this also requires a lot of effort to keep them in sync.

zero323 replied:
Pre 3.6 wasn't an issue anyway.. In practice, given the size of the project and overall complexity of annotations, stubs seem to be a somewhat better choice anyway (avoid cyclic imports or certain `Generic` pitfalls for starters, not being stuck in legacy Python support). This means that potential migration might be relatively easy.

My biggest concern is that over four years I haven't seen any active Spark contributors willing to help with maintenance (and that's despite gentle  pings). Additionally annotation ecosystem is still evolving ‒ so having a project that can iterate faster than annotated codebase is not a bad thing. 

The biggest advantage of maintaining annotations alongside actual codebase is that it requires conscious choices of signatures ‒ not every type of signature can be annotated in a meaningful way and some can result in exploding overloads. But it doesn't really remove the maintenance overhead ‒ for example changes in the ML API, mostly invisible for the final user, wreaked havoc, as it is the most vulnerable part of code (a lot of generics there). 

Anyway... That's just my 2 cents. If you reanimate the discussion. please ping me ‒ I'd like to see how it goes. TIA

Fokko replied:
I'm happy to work on the annotations, but would be great to have some support from the committers as well. I'm planning to integrate mypy in the CI as well, so we can enforce the type checking in the main repository as well. Personally, I don't see any harm in adding the types over time, and work towards coverage, starting with the public functions because they add the most value.

zero323 replied:
> Personally, I don't see any harm in adding the types over time, and work towards coverage,

The biggest problem is the resolution (https://bugs.python.org/issue39168 seems to be fixed now). As far as I am aware it is all-or-nothing at the module level (assuming `partial` mode), so working over time will likely degrade experience of the users that already use annotations.

And it seems somewhat unnecessary, considering that you could migrate stubs into main repo with a single copy, if there was a will to maintain the result. Inline variant would be trickier (and there is still concern of https://bugs.python.org/issue39168).

zero323 replied:
How about continue the discussion [here](Given a discussion related to <a href="https://github.com/apache/spark/pull/29122">SPARK-32320 PR</a> I'd like to resurrect this thread. Is there any interest in migrating annotations to the main repository?) for a better visibility?

Fokko replied:
The bug is fixed in 3.6, so that shouldn't be a problem. Better to move this to the maillist indeed, thanks!

Fokko replied:
I've removed all the inline type annotations :)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

zero323 commented on python/pyspark/ml/tuning.py:
```diff
@@ -500,13 +501,13 @@ class CrossValidatorModel(Model, _CrossValidatorParams, MLReadable, MLWritable):
     """
 
-    def __init__(self, bestModel, avgMetrics=[], subModels=None):
+    def __init__(self, bestModel, avgMetrics: List = None, subModels=None):
```
Out of curiosity ‒why do we need default here?

Fokko replied:
Good question, I don't think these should be optional

Fokko replied:
At the same time, this would change the public API

ueshin commented on python/pyspark/sql/functions.py:
```diff
@@ -2440,7 +2442,7 @@ def from_json(col, schema, options={}):
 
 @since(2.1)
-def to_json(col, options={}):
+def to_json(col, options: Dict = {}):
```
We should use `None` here as well?

Fokko committed:
[SPARK-32320][PYSPARK] Remove mutable default arguments

This is bad practice, and might lead to unexpected behaviour:
https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Add bugbear to check it in the CI

Fokko commented:
Fixed the voilations:
```
./python/pyspark/ml/regression.py:1743:40: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
                 quantileProbabilities=list([0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]),
                                       ^
./python/pyspark/ml/regression.py:1761:41: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
                  quantileProbabilities=list([0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]),
                                        ^
./python/pyspark/ml/tuning.py:511:46: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, bestModel, avgMetrics=[], subModels=None):
                                             ^
./python/pyspark/ml/tuning.py:871:53: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, bestModel, validationMetrics=[], subModels=None):
                                                    ^
./python/pyspark/resource/profile.py:35:63: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
                                                              ^
./python/pyspark/resource/profile.py:35:77: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
                                                                            ^
./python/pyspark/sql/functions.py:2479:36: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_json(col, schema, options={}):
                                   ^
./python/pyspark/sql/functions.py:2527:26: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def to_json(col, options={}):
                         ^
./python/pyspark/sql/functions.py:2567:34: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def schema_of_json(json, options={}):
                                 ^
./python/pyspark/sql/functions.py:2597:32: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def schema_of_csv(csv, options={}):
                               ^
./python/pyspark/sql/functions.py:2623:25: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def to_csv(col, options={}):
                        ^
./python/pyspark/sql/functions.py:2934:35: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_csv(col, schema, options={}):
                                  ^
./python/pyspark/sql/avro/functions.py:29:47: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_avro(data, jsonFormatSchema, options={}):
                                              ^
./dev/sparktestsupport/modules.py:34:97: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, name, dependencies, source_file_regexes, build_profile_flags=(), environ={},
                                                                                                ^
14    B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
14
1
```

Furthermore:
- Added `flake8-bugbear` to the CI. This is a well known additional set of rules next to flake8: https://github.com/pycqa/flake8-bugbear
- Added the `B006` rule, which is the mutable arguments rule :)

zero323 commented on python/pyspark/sql/avro/functions.py:
```diff
@@ -26,7 +26,7 @@
 
 @since(3.0)
-def from_avro(data, jsonFormatSchema, options={}):
+def from_avro(data, jsonFormatSchema, options=None):
```
We should change annotations from

```python
def from_avro(
    data: ColumnOrName, jsonFormatSchema: str, options: Dict[str, str] = ...
) -> Column: ..
```

to

```python
def from_avro(
    data: ColumnOrName, jsonFormatSchema: str, options: Optional[Dict[str, str]] = ...
) -> Column: ..
```

Fokko replied:
Thanks. Just curious, how can we check this? Mypy has been added, and I would expect that this would be caught.

zero323 replied:
>  Mypy has been added, and I would expect that this would be caught.

As far as I recall this behavior depends on the MyPy flags. In particular see [Disabling strict optional checking](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#no-strict-optional).

So adding `no_implicit_optional` to `mypy.ini` should catch this.

zero323 replied:
FYI I am trying to establish optimal set of flags for mypy config, but it is a bit slowish work ‒ unlike standalone stubs, we have to correct for internal modules, tests and such, and I'd prefer to avoid atomic options (ignoring all errors there).

Fokko replied:
Thanks! 👍  Yes, definitely a process that we have to go through. Let's break this up into smaller pieces.

zero323 commented on python/pyspark/sql/functions.py:
```diff
@@ -2476,7 +2478,7 @@ def json_tuple(col, *fields):
 
 @since(2.1)
-def from_json(col, schema, options={}):
+def from_json(col, schema, options=None):
```
Same as for `sql.avro.functions`, we'll need here `Optional[Dict[str, str]]` (applies also to the remaining functions).

zero323 commented on python/pyspark/ml/tuning.py:
```diff
@@ -508,13 +508,13 @@ class CrossValidatorModel(Model, _CrossValidatorParams, MLReadable, MLWritable):
     """
 
-    def __init__(self, bestModel, avgMetrics=[], subModels=None):
+    def __init__(self, bestModel, avgMetrics=None, subModels=None):
```
Annotation

https://github.com/apache/spark/blob/46ad325e56abd95c0ffdbe64aad78582da8c725d/python/pyspark/ml/tuning.pyi#L107

should change to  `avgMetrics: Optional[List[float]] = ...,`

zero323 commented on python/pyspark/ml/tuning.py:
```diff
@@ -868,12 +868,12 @@ class TrainValidationSplitModel(Model, _TrainValidationSplitParams, MLReadable,
     """
 
-    def __init__(self, bestModel, validationMetrics=[], subModels=None):
+    def __init__(self, bestModel, validationMetrics=None, subModels=None):
```
Annotation

https://github.com/apache/spark/blob/46ad325e56abd95c0ffdbe64aad78582da8c725d/python/pyspark/ml/tuning.pyi#L174

should change to `validationMetrics: Optional[List[float]] = ...,`

zero323 commented on python/pyspark/ml/regression.py:
```diff
@@ -1740,7 +1742,7 @@ class AFTSurvivalRegression(_JavaRegressor, _AFTSurvivalRegressionParams,
     def __init__(self, *, featuresCol="features", labelCol="label", predictionCol="prediction",
                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
-                 quantileProbabilities=list([0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]),
+                 quantileProbabilities=DEFAULT_QUANTILE_PROBABILITIES,
```
This is still a mutable default, isn't it? Shouldn't `DEFAULT_QUANTILE_PROBABILITIES` by a `tuple`, if we want to be consistent?

Fokko replied:
Bugbear doesn't complain, but you're right:
```
MacBook-Pro-van-Fokko:spark fokkodriesprong$ python3
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 26 2018, 19:50:54) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import List
>>> 
>>> DEFAULT_QUANTILE_PROBABILITIES = [0.01, 0.1, 0.99]
>>> 
>>> def mutate_arguments(mutate: List = DEFAULT_QUANTILE_PROBABILITIES):
...   mutate.append(22)
...   print(mutate)
... 
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22, 22, 22]
```

zero323 commented on python/pyspark/ml/regression.py:
```diff
@@ -1654,7 +1656,7 @@ class _AFTSurvivalRegressionParams(_PredictorParams, HasMaxIter, HasTol, HasFitI
         super(_AFTSurvivalRegressionParams, self).__init__(*args)
         self._setDefault(censorCol="censor",
-                         quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
+                         quantileProbabilities=DEFAULT_QUANTILE_PROBABILITIES,
```
Side note ‒ I remember this was included for Scala parity (SPARK-32310), but isn't that a dead code, give we use the same default in estimator and model constructors?

Fokko replied:
Yes, you are correct. The code isn't being used. I've removed the default argument.

zero323 replied:
Sorry for the confusion ‒ I just wanted to point it out to see if I missed something, but I don't think we should change it here, as this will be a breaking change.

Fokko replied:
I've removed the default argument in the dead code.

Fokko replied:
Yes, since there is only one place where this is used, let's revert this.

Fokko committed:
Update the annotations

Fokko committed:
Remove unused DEFAULT_QUANTILE_PROBABILITIES

HyukjinKwon commented:
ok to test

HyukjinKwon commented:
cc @huaxingao and @zhengruifeng FYI

SparkQA commented:
**[Test build #130061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130061/testReport)** for PR 29122 at commit [`e699c49`](https://github.com/apache/spark/commit/e699c49f171def64bf96b14386e610826c1413aa).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34670/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34670/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34670/
Test FAILed.

SparkQA commented:
**[Test build #130061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130061/testReport)** for PR 29122 at commit [`e699c49`](https://github.com/apache/spark/commit/e699c49f171def64bf96b14386e610826c1413aa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130061/
Test FAILed.

Fokko committed:
Revert minor changes

SparkQA commented:
**[Test build #130101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130101/testReport)** for PR 29122 at commit [`ee7a763`](https://github.com/apache/spark/commit/ee7a763ac0e4704830fbf7d2396f88ccc89592fe).

Fokko committed:
Disallow implicit optional

SparkQA commented:
**[Test build #130103 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130103/testReport)** for PR 29122 at commit [`1da61d4`](https://github.com/apache/spark/commit/1da61d45d12b60d883e934d4c7fbc1232123bc20).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34710/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34710/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34710/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34712/

Fokko committed:
Ignore the deprecated API

SparkQA commented:
**[Test build #130104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130104/testReport)** for PR 29122 at commit [`cb04d54`](https://github.com/apache/spark/commit/cb04d54909b6bd4c2cf44e4a19dd5e71aec422b6).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34712/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34712/
Test PASSed.

SparkQA commented:
**[Test build #130101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130101/testReport)** for PR 29122 at commit [`ee7a763`](https://github.com/apache/spark/commit/ee7a763ac0e4704830fbf7d2396f88ccc89592fe).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130101/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34713/

SparkQA commented:
**[Test build #130103 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130103/testReport)** for PR 29122 at commit [`1da61d4`](https://github.com/apache/spark/commit/1da61d45d12b60d883e934d4c7fbc1232123bc20).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130103/
Test FAILed.

Fokko committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-32320

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34713/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34713/
Test PASSed.

SparkQA commented:
**[Test build #130107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130107/testReport)** for PR 29122 at commit [`e011266`](https://github.com/apache/spark/commit/e011266fa56f5bb652569576b45be933650c488b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34716/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34716/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34716/
Test FAILed.

SparkQA commented:
**[Test build #130104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130104/testReport)** for PR 29122 at commit [`cb04d54`](https://github.com/apache/spark/commit/cb04d54909b6bd4c2cf44e4a19dd5e71aec422b6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130104/
Test PASSed.

SparkQA commented:
**[Test build #130107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130107/testReport)** for PR 29122 at commit [`e011266`](https://github.com/apache/spark/commit/e011266fa56f5bb652569576b45be933650c488b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130107/
Test PASSed.

Fokko committed:
Merge branch 'master' into SPARK-32320

SparkQA commented:
**[Test build #131539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131539/testReport)** for PR 29122 at commit [`1b5a7aa`](https://github.com/apache/spark/commit/1b5a7aab7d71257ecaa609392150bd7803104075).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36142/

SparkQA commented:
**[Test build #131539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131539/testReport)** for PR 29122 at commit [`1b5a7aa`](https://github.com/apache/spark/commit/1b5a7aab7d71257ecaa609392150bd7803104075).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131539/
Test FAILed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36142/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36142/
Test PASSed.

zero323 commented on python/pyspark/sql/functions.py:
```diff
@@ -3471,7 +3473,7 @@ def from_json(col, schema, options={}):
 
 
-def to_json(col, options={}):
+def to_json(col, options=None):
```
Seems like we still have to modify a few annotations, right?

Probably something like [functions.pyi.patch.txt](https://github.com/apache/spark/files/5583207/functions.pyi.patch.txt)

Fokko replied:
Good catch, I've added them

Fokko replied:
Not sure why implicit optional is still allowed 🤔

SparkQA commented:
**[Test build #131606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131606/testReport)** for PR 29122 at commit [`1b5a7aa`](https://github.com/apache/spark/commit/1b5a7aab7d71257ecaa609392150bd7803104075).

SparkQA commented:
**[Test build #131606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131606/testReport)** for PR 29122 at commit [`1b5a7aa`](https://github.com/apache/spark/commit/1b5a7aab7d71257ecaa609392150bd7803104075).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131606/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36207/

Fokko committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-32320

Fokko committed:
Add explicit optionals

SparkQA commented:
**[Test build #131658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131658/testReport)** for PR 29122 at commit [`0e372ca`](https://github.com/apache/spark/commit/0e372caa1d3ea33ebdde98de3b4a1afbb4a5fc38).

SparkQA commented:
**[Test build #131658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131658/testReport)** for PR 29122 at commit [`0e372ca`](https://github.com/apache/spark/commit/0e372caa1d3ea33ebdde98de3b4a1afbb4a5fc38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131658/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36259/

Fokko committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-32320

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131700/

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #131701 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131701/testReport)** for PR 29122 at commit [`b628156`](https://github.com/apache/spark/commit/b6281564c198dd3690cc9e6a7a769dfbff5ebe57).

SparkQA commented:
**[Test build #131701 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131701/testReport)** for PR 29122 at commit [`b628156`](https://github.com/apache/spark/commit/b6281564c198dd3690cc9e6a7a769dfbff5ebe57).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131701/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36299/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36300/

Fokko committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-32320

SparkQA commented:
**[Test build #131765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131765/testReport)** for PR 29122 at commit [`c75cd57`](https://github.com/apache/spark/commit/c75cd574abfbaa6b9f5f99079a1ba25418d55720).

SparkQA commented:
**[Test build #131765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131765/testReport)** for PR 29122 at commit [`c75cd57`](https://github.com/apache/spark/commit/c75cd574abfbaa6b9f5f99079a1ba25418d55720).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowColumns(`

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131765/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36362/

zhengruifeng reviewed (approved):
@Fokko please update the description of this PR.

Fokko committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-32320

Fokko commented:
@zhengruifeng done! :)

SparkQA commented:
**[Test build #132225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132225/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36825/

SparkQA commented:
**[Test build #132225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132225/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132225/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36825/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36825/

HyukjinKwon commented:
Just for the record and FWIW I don't mind merging this PR. I will leave it to other committers here.

zhengruifeng commented:
retest this please

SparkQA commented:
**[Test build #132313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132313/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36913/

SparkQA commented:
**[Test build #132313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132313/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132313/

zhengruifeng commented:
retest this please

SparkQA commented:
**[Test build #132338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132338/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36938/

SparkQA commented:
**[Test build #132338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132338/testReport)** for PR 29122 at commit [`11f3790`](https://github.com/apache/spark/commit/11f3790ac3ccc15efd7944e0fc1a3a2e3e8e3247).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132338/

HyukjinKwon commented:
@zhengruifeng the tests are broken globally for an unknown reason. It's being investigated at https://github.com/apache/spark/pull/30645 and it's hardly related to the current change. I think it's fine to merge if the test results bother you.

zhengruifeng commented:
@HyukjinKwon Thanks for your explanation!

zhengruifeng commented:
Merged to master, thanks all!
